### ISSUE TITLE
Error executing 'postInstallation: ELOOP: too many symbolic links encountered, lstat /opt/bitnami/redmine/conf/database.yml'

### ISSUE THREAD

**Issue by ToniDS** [bitnami/bitnami-docker-redmine#113]
**Description**

After restarting my server, I cannot get redmine to work anymore. I have a reverse proxy in my host, which worked perfectly before. Now, the first error I received was that passenger could not bind on port 3000 anymore, as port 3000 is already in use. I then tried updating the images (with docker pull), and now, I get the error described abpve. Postgresql shows: database system is ready to accept connections, but redmine doesn't come up. 

**Steps to reproduce the issue:**

1. Restart server 
2. automatic restart
3. docker-compose -f docker-compose-postgresql.yml up

**Describe the results you received:**

Error: Error executing 'postInstallation: ELOOP: too many symbolic links encountered, lstat /opt/bitnami/redmine/conf/database.yml'

**Describe the results you expected:**
Running postgresql and redmine container, accessible via port 3000 on loopback

**Additional information you deem important (e.g. issue happens only occasionally):**

This is my docker compose: 
```

version: '2'

services:
  postgresql:
    image: 'bitnami/postgresql:latest'
    volumes:
      - 'postgresql_data:/bitnami'
  redmine:
    image: 'bitnami/redmine:latest'
    ports:
      - '127.0.0.1:1025:3000'
    environment:
      - REDMINE_DB_POSTGRES=postgresql
    volumes:
      - 'redmine_data:/bitnami'
    depends_on:
      - postgresql
volumes:
  postgresql_data:
    driver: local
  redmine_data:
    driver: local
```

I had restart: always on here for both containers. 

**Version**

- Output of `docker version`:

```
Client:
 Version:           18.09.1-ce
 API version:       1.39
 Go version:        go1.11.4
 Git commit:        4c52b901c6
 Built:             Thu Jan 10 06:51:04 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.1-ce
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.4
  Git commit:       4c52b901c6
  Built:            Thu Jan 10 06:50:46 2019
  OS/Arch:          linux/amd64
  Experimental:     false

```

- Output of `docker info`:

```
Containers: 4
 Running: 1
 Paused: 0
 Stopped: 3
Images: 7
Server Version: 18.09.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9f2e07b1fc1342d1c48fe4d7bbb94cb6d1bf278b.m
runc version: 079817cc26ec5292ac375bb9f47f373d33574949
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.20.6-arch1-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 62.79GiB
Name: arch.rlwinm.de
ID: OFDO:Y44N:43SE:CJOP:3QXZ:7JOP:BAHF:2MBC:4WTW:NRO3:7U2E:6W72
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.23.2, build unknown
docker-py version: 3.7.0
CPython version: 3.7.2
OpenSSL version: OpenSSL 1.1.1a  20 Nov 2018
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
Host: arch

**Comment by ToniDS** [comment]
Upon trying to rerun and reproduce this error, now I get this: 

``
redmine_1     | postgre INFO  PostgreSQL server listening and working at postgresql:5432
redmine_1     | redmine INFO  Generating new secret...
redmine_1     | Error executing 'postInstallation': bundler: failed to load command: rake (/opt/bitnami/redmine/vendor/b
redmine_1     | Bundler::ProductionError: You are trying to install in deployment mode after changing
redmine_1     | your Gemfile. Run `bundle install` elsewhere and add the
redmine_1     | updated Gemfile.lock to version control.
redmine_1     | 
redmine_1     | If this is a development machine, remove the /opt/bitnami/redmine/Gemfile freeze 
redmine_1     | by running `bundle install --no-deployment`.
redmine_1     | 
redmine_1     | The dependencies in your gemfile changed
redmine_1     | 
redmine_1     | You have added to the Gemfile:
redmine_1     | * redmine_crm
redmine_1     | * redmine_extensions
redmine_1     | 
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/definition.rb:466:in `ensure_equi
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:13:in `setup'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler.rb:107:in `setup'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/setup.rb:20:in `<top (required)>'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
redmine_1     | 
toni_redmine_1 exited with code 1
``

**Comment by marcosbc** [comment]
Hi @ToniDS, keep in mind that if you don't delete the container, it will contain the same files after rerun.

Since the required files for persistence are already being copied to the Docker volume `redmine_data`, you should be fine by removing the containers and starting again:

```
$ docker-compose down
$ docker-compose up
```

After that you should not find the error anymore.

**Comment by ToniDS** [comment]
Thanks for your comment. Unfortunately, this is the error I get: 
`postgresql_1  | 2019-02-12 19:14:49.632 GMT [89] LOG:  database system is ready to accept connections
redmine_1     | postgre INFO  PostgreSQL server listening and working at postgresql:5432
redmine_1     | redmine INFO  Generating new secret...
redmine_1     | Error executing 'postInstallation': bundler: failed to load command: rake (/opt/bitnami/redmine/vendor/bundle/ruby/2.4.0/bin/rake)
redmine_1     | Bundler::ProductionError: You are trying to install in deployment mode after changing
redmine_1     | your Gemfile. Run `bundle install` elsewhere and add the
redmine_1     | updated Gemfile.lock to version control.
redmine_1     | 
redmine_1     | If this is a development machine, remove the /opt/bitnami/redmine/Gemfile freeze 
redmine_1     | by running `bundle install --no-deployment`.
redmine_1     | 
redmine_1     | The dependencies in your gemfile changed
redmine_1     | 
redmine_1     | You have added to the Gemfile:
redmine_1     | * redmine_crm
redmine_1     | 
redmine_1     | You have deleted from the Gemfile:
redmine_1     | * mysql2 (~> 0.5.0)
redmine_1     | 
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/definition.rb:466:in `ensure_equivalent_gemfile_and_lockfile'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:13:in `setup'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler.rb:107:in `setup'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/gems/2.4.0/gems/bundler-2.0.1/lib/bundler/setup.rb:20:in `<top (required)>'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
redmine_1     |   /opt/bitnami/ruby/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
redmine_1     | 
`
I did not add or delete anything from redmine, and I cannot run "bundle install --no-deployment", becaus the container fails to come up long enough for me to run anything in the container.

**Comment by marcosbc** [comment]
Hi @ToniDS, could it be possible you installed a plugin that requires Gems?

Note that you can temporarily start your Redmine container adding these properties to your `docker-compose.yml` file (below the `redmine:` line):

```
entrypoint: tail
command: ['-f', '/dev/null']
```

Then, enter the container with Bash and run the command below to start Redmine:

    /app-entrypoint.sh /run.sh

And when it fails, debug your application in case you missed a dependency or something like that. In that case you may need to create a custom Dockerfile to install them.

**Comment by ToniDS** [comment]
Thanks a lot! I managed to recreate the redmine instance using pg_dump/restore and copying the files folder over. I have indeed installed a plugin that requires additional gems. Do I need to add these additional gems to the dockerfile from now on to make sure I can restart my server? 
Where can I find the dockerfile to edit?

**Comment by freefly42** [comment]
The root cause of this is all the symbolic linking set up in the Bitnami init process--the configuration.yml in /bitnami/redmine/conf links back to the symlinks in /opt/bitnami/redmine, and you get a circular reference unless there is a volume mounted to cover /bitnami.  But this means it can't run by default if no volume is mounted and it seems like it would be better if the /bitnami files weren't symlinks as it would give some flexibility as to what to keep autogenerated vs retained in a volume.

**Comment by marcosbc** [comment]
Hi @freefly42, thanks for your feedback. Check our comments below:

>The root cause of this is all the symbolic linking set up in the Bitnami init process--the configuration.yml in /bitnami/redmine/conf links back to the symlinks in /opt/bitnami/redmine, and you get a circular reference unless there is a volume mounted to cover /bitnami.

Note that the issues you mention are happen when a container initialization goes wrong and does not complete successfully. If no volume is mounted, then the initialization will start from zero (thinking the installation is new, without any existing data).

> But this means it can't run by default if no volume is mounted and it seems like it would be better if the /bitnami files weren't symlinks as it would give some flexibility as to what to keep autogenerated vs retained in a volume.

Note that an initialization where a volume is not mounted does not fail like you mention. The only thing is that when you do "docker-compose down && docker-compose up" you would have lost your data.
______

Note that the reason why we use symlinks originate from application-specific implementation issues, as they were not implemented with a "cloud-native" approach:

* Some applications don't have a clear logic-data separation. I.e. there are .rb files in a directory with site-generated content
* Some others do, but force you to use hardcoded paths inside the main application directory. I.e. it is not possible to change the "files" directory location without modifying the application source code

To solve this, we copy what we consider "data" (i.e. Redmine "files" directory) to the volume, then generate a symlink during the initialization. We have found the symlinks approach the easiest to implement and the ones that generated the less issues for users.

Another approach would be to persist the entire "htdocs" directory, but that would make it impossible to automatically upgrade the Redmine version by changing the container image (which in some cases may see like a natural thing to do).

**Comment by freefly42** [comment]
Correct, but I want to remove the symlink for the config because everything that's in it is generated from my docker-compose and I have a volume just for the preserved data (I have plugins installed in the image).  Not sure of the point of the symlink back to the original symlink instead of an empty folder or an empty file since that symlink can never point to a valid file.

**Comment by marcosbc** [comment]
> Correct, but I want to remove the symlink for the config because everything that's in it is generated from my docker-compose and I have a volume just for the preserved data (I have plugins installed in the image). Not sure of the point of the symlink back to the original symlink instead of an empty folder or an empty file since that symlink can never point to a valid file.

That use case is currently not supported. In any case, it should be as simple for you as to copy the configuration file to the existing volume. We are evaluating some longer term improvements to the way data is persisted in Bitnami containers, but it may take some time to get implemented. For now, I'm going to create an internal task for not persisting Redmine configuration files, as they don't contain any data that cannot be generated between runs.

In the meantime, if you want to avoid persisting the config file on your side, you can fork the container image and edit the following file:

```
/root/.nami/components/com.bitnami.redmine/nami.json
```

The following section would need to be modified:

```
        "dataToPersist": {"value":         [
            "{{$app.confDir}}/configuration.yml",
            "{{$app.confDir}}/database.yml",
            "public/plugin_assets",
            "plugins",
            "files"
        ]},
```

So if you remove the configuration.yml/database.yml lines, create that image and run it, you should notice it no longers uses those persisted files.

**Comment by vadonka** [comment]
I have the same issue, its happening even with a fresh deployment with empty volume. It was working since i last pull the fresh image. Something rigged in the image what cause this.

**Comment by marcosbc** [comment]
> I have the same issue, its happening even with a fresh deployment with empty volume. It was working since i last pull the fresh image. Something rigged in the image what cause this.

Could you share the steps so we can reproduce this on our side? The more info the better (i.e. docker-compose file, environment configurations, etc.)

**Comment by LeeThompson** [comment]
I just got this on a fresh pull (tonight).

Docker version 18.09.8, build bfed4f5 (Synology)
Bitnami Image Version: 4.1.1-debian-10-r104

Docker on the Synology has a GUI, I have no idea if it's using docker-compose behind the scenes.

**Comment by LeeThompson** [comment]
I tried copying a dummy (blank) configuration.xml over the one in the container and got this:

```
touch configuration.xml
docker cp configuration.xml Redmine:/opt/bitnami/redmine/conf/configuration.yml
```

`Error response from daemon: evalSymlinksInScope: too many links in /volume1/@docker/btrfs/subvolumes/96a623ca71dcfc1963f8a20b30c311afdd9f640654b3eaa8aede3b38b07a97dd/opt/bitnami/redmine/conf`

So it's something that happens when the container initializes.

**Comment by marcosbc** [comment]
The current Redmine container does not support mounting `configuration.yml` nor `database.yml` since these are renamed to the volume, and then a symlink to them is created.

If you need to apply a custom configuration file, we recommend you to let the container initialize for the first time and then apply the changes to the configuration file, in the volume (i.e. replacing it).

If that's not an option and you need the file to be modified during the first deployment, then you would need to workaround the image's [`app-entrypoint.sh` entrypoint](https://github.com/bitnami/bitnami-docker-redmine/blob/master/4/debian-10/rootfs/app-entrypoint.sh) and their create a new file, or modify it by extending the image after the first initialization completed, i.e..

```diff
     . /init.sh
     nami_initialize redmine
     info "Starting gosu... "
+    cp /mounted/configuration.yml /bitnami/redmine/conf/configuration.yml
 fi

 exec tini -- "$@"
```

**Comment by LeeThompson** [comment]
@marcosbc I didn't change or do anything, this is stock.

**Comment by marcosbc** [comment]
Sorry for my confusion. I think I understand now: You first obtained the error, then copied the file to try and fix it, but get the same error. 

The error usually happens when the first initialization of the Redmine container fails. If your deployment does not contain any important data, could you try and recreate it by removing the volumes and starting again?

If you were not using Docker volumes, but the data is mounted from a sub-folder, then delete the related folders for the deployment to succeed. You should see this when starting the deployment:

```
redmine_1  | redmine INFO  Configuring Redmine database...
```

But NOT this:

```
redmine_1  | redmine INFO  Redmine has been already initialized, restoring...
```

**Comment by marcosbc** [comment]
Hi, FYI we recently released a new container version, [4.2.1-debian-10-r70](https://github.com/bitnami/bitnami-docker-redmine#421-debian-10-r70). The initialization logic has received a major refactor, and its scripts are now located in the [`rootfs/` directory](https://github.com/bitnami/bitnami-docker-redmine/blob/master/4/debian-10/rootfs/).

As part of these changes, the `ELOOP: too many symbolic links encountered` issue should not happen anymore, so I'm proceeding to close this issue.
